library/Booster/JsonRpc.hs

```diff
@@ -155,14 +159,40 @@ respond stateVar =
                             unless (null newModule.symbols) $
                                 throwE . AddModuleError $
                                     "Module introduces new symbols: " <> listNames newModule.symbols
-                     in case runExcept (checkModule >> addToDefinitions newModule state.definitions) of
+
+                            when nameAsId $
+                                case (Map.lookup (getId $ newModule.name) state.definitions, Map.lookup moduleHash state.definitions) of
+                                    (Just{}, Nothing) ->
+                                        -- another module with the same name already exists
+                                        throwE . AddModuleError $ "Duplicate Module Name"
+                                    (Just nmMod, Just idMod)
+                                        | nmMod /= idMod ->
+                                            -- this module has previously been added and different
+                                            -- module with the same name also already exists
+                                            throwE . AddModuleError $ "Duplicate Module Name"
+                                        | otherwise ->
+                                            -- this module has previously been added with name-as-id: true
+                                            -- we can allow this, since the contents of the named module
+                                            -- are the same
+                                            pure ()
+                                    _ -> pure ()
+                     in case runExcept
+                            (checkModule >> addToDefinitions newModule{ParsedModule.name = Id moduleHash} state.definitions) of
                             Left err ->
                                 abortWith $ RpcError.backendError RpcError.CouldNotVerifyPattern err
```

Comment:
We should use the new `DuplicateModuleName` here. Which means 
1. a error type other than `AddModuleError` to throw above (probably just `DuplicateModule`)
2. a separate match to return `DuplicateModuleName` here
```suggestion
                            Left (DuplicateModule name) ->
                                abortWith $ RpcError.backendError RpcError.DuplicateModuleName name
                            Left err ->
                                abortWith $ RpcError.backendError RpcError.CouldNotVerifyPattern err
```